File: panel/backend/internal/models/notification.go
package models

import "time"

// NotificationPriority is the urgency of a notification.
type NotificationPriority string

const (
	NotificationPriorityLow      NotificationPriority = "low"
	NotificationPriorityNormal   NotificationPriority = "normal"
	NotificationPriorityHigh     NotificationPriority = "high"
	NotificationPriorityUrgent   NotificationPriority = "urgent"
	NotificationPriorityCritical NotificationPriority = "critical"
)

// NotificationCategory groups notifications the way ActivityCategory does.
type NotificationCategory string

const (
	NotificationCategorySystem      NotificationCategory = "system"
	NotificationCategoryUser        NotificationCategory = "user"
	NotificationCategoryRole        NotificationCategory = "role"
	NotificationCategoryNode        NotificationCategory = "node"
	NotificationCategoryTemplate    NotificationCategory = "template"
	NotificationCategoryInstance    NotificationCategory = "instance"
	NotificationCategoryAPIKey      NotificationCategory = "api_key"
	NotificationCategoryAuth        NotificationCategory = "auth"
	NotificationCategoryMod         NotificationCategory = "mod"
	NotificationCategoryApplication NotificationCategory = "application"
	NotificationCategorySecurity    NotificationCategory = "security"
	NotificationCategoryTheme       NotificationCategory = "theme"
	NotificationCategoryUpdate      NotificationCategory = "update"
	NotificationCategoryGeneral     NotificationCategory = "general"
)

// AllNotificationCategories is the whitelist used for validation server-side.
var AllNotificationCategories = []NotificationCategory{
	NotificationCategorySystem,
	NotificationCategoryUser,
	NotificationCategoryRole,
	NotificationCategoryNode,
	NotificationCategoryTemplate,
	NotificationCategoryInstance,
	NotificationCategoryAPIKey,
	NotificationCategoryAuth,
	NotificationCategoryMod,
	NotificationCategoryApplication,
	NotificationCategorySecurity,
	NotificationCategoryTheme,
	NotificationCategoryUpdate,
	NotificationCategoryGeneral,
}

// AllNotificationPriorities is the whitelist used for validation.
var AllNotificationPriorities = []NotificationPriority{
	NotificationPriorityLow,
	NotificationPriorityNormal,
	NotificationPriorityHigh,
	NotificationPriorityUrgent,
	NotificationPriorityCritical,
}

// Notification is one inbox row for a specific user. Broadcasts are stored as
// one row per recipient (fan-out at create time) so per-user read/delete stays
// simple and concurrent — no "read by user X masks read by user Y" bug.
type Notification struct {
	ID         int64                `json:"id"`
	UserID     int64                `json:"user_id"`
	ActorID    *int64               `json:"actor_id,omitempty"`
	ActorName  string               `json:"actor_name"`
	Category   NotificationCategory `json:"category"`
	Priority   NotificationPriority `json:"priority"`
	Title      string               `json:"title"`
	Message    string               `json:"message"`
	Link       string               `json:"link,omitempty"`
	ActionLabel string              `json:"action_label,omitempty"`
	Metadata   string               `json:"metadata,omitempty"`
	IsRead     bool                 `json:"is_read"`
	IsBroadcast bool                `json:"is_broadcast"`
	CreatedAt  time.Time            `json:"created_at"`
	ReadAt     *time.Time           `json:"read_at,omitempty"`
}

// NotificationMode controls how a user receives notifications:
// realtime = WS push + immediate email, digest = WS push + daily email
// summary, off = inbox only (no push, no email).
type NotificationMode string

const (
	NotificationModeRealtime NotificationMode = "realtime"
	NotificationModeDigest   NotificationMode = "digest"
	NotificationModeOff      NotificationMode = "off"
)

// ValidNotificationModes is the whitelist for prefs validation.
var ValidNotificationModes = map[string]bool{
	"realtime": true, "digest": true, "off": true,
}

// NotificationPrefs is one user's 065 delivery preferences.
type NotificationPrefs struct {
	UserID       int64            `json:"user_id"`
	Mode         NotificationMode `json:"mode"`
	EmailOptOut  bool             `json:"email_opt_out"`
	LastDigestAt *time.Time       `json:"last_digest_at,omitempty"`
}

File: panel/frontend/src/features/notifications/types/notification.ts
export type NotificationCategory =
  | 'system'
  | 'user'
  | 'role'
  | 'node'
  | 'template'
  | 'instance'
  | 'api_key'
  | 'auth'
  | 'mod'
  | 'application'
  | 'security'
  | 'theme'
  | 'update'
  | 'general';

export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent' | 'critical';

export interface Notification {
  id: number;
  user_id: number;
  actor_id?: number;
  actor_name: string;
  category: NotificationCategory;
  priority: NotificationPriority;
  title: string;
  message: string;
  link?: string;
  action_label?: string;
  metadata?: string;
  is_read: boolean;
  is_broadcast: boolean;
  created_at: string;
  read_at?: string;
}

export interface NotificationStats {
  total: number;
  unread: number;
  by_category: Record<string, number>;
  by_priority: Record<string, number>;
  broadcast?: number;
  read?: number;
}

export interface CreateNotificationPayload {
  user_id?: number;
  category?: NotificationCategory;
  priority?: NotificationPriority;
  title: string;
  message: string;
  link?: string;
  action_label?: string;
  metadata?: string;
  broadcast?: boolean;
}

export const CATEGORY_META: Record<NotificationCategory, { label: string; icon: string; color: string }> = {
  system:      { label: 'System',      icon: 'system',      color: '#38bdf8' },
  user:        { label: 'User',        icon: 'users',       color: '#a78bfa' },
  role:        { label: 'Role',        icon: 'shield',      color: '#f472b6' },
  node:        { label: 'Node',        icon: 'nodes',       color: '#34d399' },
  template:    { label: 'Template',    icon: 'templates',   color: '#60a5fa' },
  instance:    { label: 'Instance',    icon: 'instances',   color: '#fbbf24' },
  api_key:     { label: 'API Key',     icon: 'apikeys',     color: '#fb923c' },
  auth:        { label: 'Auth',        icon: 'auth',        color: '#f87171' },
  mod:         { label: 'Mod',         icon: 'mods',        color: '#c084fc' },
  application: { label: 'Application', icon: 'apps',        color: '#2dd4bf' },
  security:    { label: 'Security',    icon: 'security',    color: '#ef4444' },
  theme:       { label: 'Theme',       icon: 'themes',      color: '#e879f9' },
  update:      { label: 'Update',      icon: 'update',      color: '#22d3ee' },
  general:     { label: 'General',     icon: 'general',     color: '#9ca3af' },
};

export const PRIORITY_META: Record<NotificationPriority, { label: string; color: string; bg: string; dot: string }> = {
  low:      { label: 'Low',      color: 'text-gray-400',  bg: 'bg-gray-700/30 border-gray-600/30',  dot: 'bg-gray-500' },
  normal:   { label: 'Normal',   color: 'text-sky-300',   bg: 'bg-sky-500/10 border-sky-400/20',    dot: 'bg-sky-400' },
  high:     { label: 'High',     color: 'text-amber-300', bg: 'bg-amber-500/15 border-amber-400/30',dot: 'bg-amber-400' },
  urgent:   { label: 'Urgent',   color: 'text-orange-300',bg: 'bg-orange-500/15 border-orange-400/30',dot: 'bg-orange-400' },
  critical: { label: 'Critical', color: 'text-red-300',   bg: 'bg-red-500/15 border-red-400/30',    dot: 'bg-red-500' },
};
